lib/elixir/lib/calendar/datetime.ex

```diff
@@ -930,11 +931,31 @@ defmodule DateTime do
       iex> DateTime.to_iso8601(dt, :basic)
       "20000229T230007-0400"
 
+      iex> dt = %DateTime{year: 2000, month: 2, day: 29, zone_abbr: "AMT",
+      ...>                hour: 23, minute: 0, second: 7, microsecond: {0, 0},
+      ...>                utc_offset: -14400, std_offset: 0, time_zone: "America/Manaus"}
+      iex> DateTime.to_iso8601(dt, :extended, 3600)
+      "2000-03-01T04:00:07+01:00"
+
+      iex> dt = %DateTime{year: 2000, month: 2, day: 29, zone_abbr: "AMT",
+      ...>                hour: 23, minute: 0, second: 7, microsecond: {0, 0},
+      ...>                utc_offset: -14400, std_offset: 0, time_zone: "America/Manaus"}
+      iex> DateTime.to_iso8601(dt, :extended, 0)
+      "2000-03-01T03:00:07+00:00"
```

Comment:
So my comment was about this test. :) The `offset` here shouldn't be zero, it should be -14400.

Reply:
To be clear, I was thinking the logic would be largely the logic we have today. The only difference is that we will add the current offset, something like this:

```diff
diff --git a/lib/elixir/lib/calendar/datetime.ex b/lib/elixir/lib/calendar/datetime.ex
index 512710635..9c2fb0c5a 100644
--- a/lib/elixir/lib/calendar/datetime.ex
+++ b/lib/elixir/lib/calendar/datetime.ex
@@ -932,9 +932,9 @@ defmodule DateTime do
 
   """
   @spec to_iso8601(Calendar.datetime(), :extended | :basic) :: String.t()
-  def to_iso8601(datetime, format \\ :extended)
+  def to_iso8601(datetime, format \\ :extended, offset \\ 0)
 
-  def to_iso8601(%{calendar: Calendar.ISO} = datetime, format)
+  def to_iso8601(%{calendar: Calendar.ISO} = datetime, format, ofset)
       when format in [:extended, :basic] do
     %{
       year: year,
@@ -952,13 +952,13 @@ defmodule DateTime do
     Calendar.ISO.date_to_string(year, month, day, format) <>
       "T" <>
       Calendar.ISO.time_to_string(hour, minute, second, microsecond, format) <>
-      Calendar.ISO.offset_to_string(utc_offset, std_offset, time_zone, format)
+      Calendar.ISO.offset_to_string(utc_offset + offset, std_offset, time_zone, format)
   end
 
-  def to_iso8601(%{calendar: _} = datetime, format) when format in [:extended, :basic] do
+  def to_iso8601(%{calendar: _} = datetime, format, offset) when format in [:extended, :basic] do
     datetime
     |> convert!(Calendar.ISO)
-    |> to_iso8601(format)
+    |> to_iso8601(format, offset)
   end
 
   @doc """
```

Is there a reason why this approach won't work? :)

Reply:
When parsing the offset is from UTC, so when applied the offset should be from UTC. This makes the most sense to me  and is the least surprising to me. Changing a timezone for the date should not affect output when I supply a specific offset, as the offset is not the timezone.